src/block/block_open.c

```diff
@@ -274,75 +276,145 @@ __wt_block_close(WT_SESSION_IMPL *session, WT_BLOCK *block)
  *	Write a file's initial descriptor structure.
  */
 int
-__wt_desc_write(WT_SESSION_IMPL *session, WT_FH *fh, uint32_t allocsize)
+__wt_desc_write(WT_SESSION_IMPL *session, WT_FH *fh,
+    uint32_t allocsize, const char *filename, const char *metadata)
 {
 	WT_BLOCK_DESC *desc;
 	WT_DECL_ITEM(buf);
 	WT_DECL_RET;
+	size_t min_config_len, total_len;
+	uint8_t *p;
+	const char *filecfg[] = { WT_CONFIG_BASE(session, file_meta), NULL };
+	char *min_config;
+
+	min_config = NULL;
 
 	/* If in-memory, we don't read or write the descriptor structure. */
 	if (F_ISSET(S2C(session), WT_CONN_IN_MEMORY))
 		return (0);
 
 	/* Use a scratch buffer to get correct alignment for direct I/O. */
-	WT_RET(__wt_scr_alloc(session, allocsize, &buf));
+	WT_ERR(__wt_scr_alloc(session, allocsize, &buf));
 	memset(buf->mem, 0, allocsize);
 
-	/*
-	 * Checksum a little-endian version of the header, and write everything
-	 * in little-endian format. The checksum is (potentially) returned in a
-	 * big-endian format, swap it into place in a separate step.
-	 */
+	/* Initialize the block descriptor structure. */
 	desc = buf->mem;
 	desc->magic = WT_BLOCK_MAGIC;
 	desc->majorv = WT_BLOCK_MAJOR_VERSION;
 	desc->minorv = WT_BLOCK_MINOR_VERSION;
 	desc->checksum = 0;
+	desc->allocsize = allocsize;
+	p = (uint8_t *)buf->mem + WT_BLOCK_DESC_SIZE;
+
+	/*
+	 * Get a copy of where the application's file configuration differs from
+	 * the base defaults. The block manager stores a minimal configuration
+	 * so we can read standalone files. Standalone files have limited space
+	 * when first created, so explicitly strip the app_metadata field since
+	 * that's the field that's likely to be large: we'll get a copy of it on
+	 * the first checkpoint (when the standalone file can grow to store as
+	 * much configuration information as we have).
+	 */
+	WT_ERR(__wt_config_discard_defaults(session,
+	    filecfg, metadata, "app_metadata=", &min_config));
```

Comment:
I'd assumed we would go the other way and only include the information necessary to decode the rest of the file in the header. Did you try that?
I'd quite like to fail creation of files with metadata greater than a certain length, probably with a flag to force creation. Which is more appealing if the header block gets a known set of configuration options.

Reply:
I didn't try that.

The difficulty is ensuring we don't skip anything we need, and of course, remembering the block manager layer when we add new items. I'm sure we could get all that right, but it's trickier. It doesn't save us on code complexity because including only the information we need doesn't ensure the information will fit, we still have to do all the work of checking because a 4KB path to a compressor will blow out the space.

An argument more to your point (because we do have to store the path to the compressor at the block layer), 4KB of column names would also blow out the space and we don't actually need column names in the block layer.

There's something else we also briefly discussed. I think it's possible to grab additional blocks at the start of the file and simply hide them from previous releases. It's going to require  a bit of magic code to pull it apart and stitch it all back together, but I think it's doable. If we did that, then this additional information is just always available, end of story. Let me take a run at that and see how deep that hole is.

Reply:
@agorrod 
> I'd assumed we would go the other way and only include the information necessary to decode the rest of the file in the header.

I'm going to have to go with your approach -- I hadn't solved the encryption problem (parts of the file configuration string require encryption, for example `columns=`), and I can't find a way to make encryption work. So, I will rework things to write only the information that's required to tear the file apart.

Reply:
That makes me a bit sad, but thanks.